src/services/codefixes/convertToAsyncFunction.ts

```diff
@@ -0,0 +1,471 @@
+            }
+        });
+        return constIdentifiers;
+    }
+
+    function findLastDotThens(func: FunctionLikeDeclaration, checker: TypeChecker): Map<boolean> {
+        if (!func.body) {
+            return createMap();
+        }
+
+        // maps nodes to boolean - true indicates that there is another .then() in the callback chain
+        const lastDotThen: Map<boolean> = createMap();
+
+        forEachChild(func.body, function visit(node: Node) {
+            const nodeType = checker.getTypeAtLocation(node);
+            if (isCallExpression(node) && nodeType && !!checker.getPromisedTypeOfPromise(nodeType) && hasPropertyAccessExpressionWithName(node, "then")) {
+                // false - there is no following .then() in the callback chain
+                lastDotThen.set(getNodeId(node).toString(), false);
+
+                forEachChild(node, function checkChildren(child: Node) {
+                    forEachChild(child, checkChildren);
+
+                    const nodeType = checker.getTypeAtLocation(child);
+                    if (isCallExpression(child) && nodeType && !!checker.getPromisedTypeOfPromise(nodeType) && hasPropertyAccessExpressionWithName(child, "then")
+                        && child.parent && !isPropertyAccessExpression(child.parent) && lastDotThen.get(getNodeId(child.parent).toString()) === false) {
+
+                        // false - there is no following .then() in the callback chain
+                        lastDotThen.set(getNodeId(child).toString(), false);
+                    }
+                    else if (isCallExpression(child) || isIdentifier(child) && nodeType && checker.getPromisedTypeOfPromise(nodeType)) {
+
+                        // true - there is a following .then() in the callback chain
+                        lastDotThen.set(getNodeId(child).toString(), true);
+                    }
+                });
+            }
+            else {
+                forEachChild(node, visit);
+            }
+        });
+
+        return lastDotThen;
+    }
+
+    function isFunctionRef(node: Node): boolean {
+        const callExpr = climbPastPropertyAccess(node);
+        return !isCallExpression(callExpr) || callExpr.expression !== node;
+    }
+
+    function declaredInFile(symbol: Symbol, sourceFile: SourceFile): boolean {
+        return symbol.valueDeclaration && symbol.valueDeclaration.getSourceFile() === sourceFile;
+    }
+
+    // varNamesMap holds all of the variables in original source code. synthNamesMap holds all of the variables created by the refactor
+    function renameCollidingVarNames(nodeToRename: FunctionLikeDeclaration, checker: TypeChecker, synthNamesMap: Map<SynthIdentifier>, context: CodeFixContextBase, lastDotThenMap: Map<boolean>, originalType: Map<Type>): FunctionLikeDeclaration {
+        const allVarNames: [Identifier, Symbol][] = [];
+
+        forEachChild(nodeToRename, function visit(node: Node) {
+            const symbol = checker.getSymbolAtLocation(node);

```

Comment:
It seems like you might want to start with `if (!isIdentifier(node)) { forEachChild(node, visit); return; }`.  That would skip a lot of semantic checks and prevent recursion within identifiers.